analysis/runtime/src/events.rs

```diff
@@ -62,7 +60,10 @@ pub enum EventKind {
     StoreAddrTaken(Pointer),
 
     /// The pointer that appears as the address result of addr_of(Local)
-    AddrOfLocal(Pointer, Local),
+    AddrOfLocal(Pointer, Local, u32),
+
+    /// The address of a constant value
+    AddrOfConst(Pointer, u32),
```

Comment:
Can you also update the comment and/or rename the variant to clarify the meaning of `AddrOfConst`?  The name is surprising to me because constants are typically rvalues, not lvalues, and thus can't have their address taken (at least not directly).  From the way it's used, it looks like this is for `Rvalue::Use(Constant)` where a constant of pointer type is copied into a local - maybe `UseConst` would be a better name?  Or `AddrOfStatic`, since the pointee must be a static allocation of some sort (either `&STATIC` or `"foo"`).

Reply:
The MIR calls them "const", e.g., `_21 = const {alloc1082: *mut *mut fdlog_st}` and `_32 = const b"/usr/local/lib\x00"` and `_63 = const "non-null function pointer"`. They are source-level constants of aggregate types, e.g., strings and arrays.

I think everything from https://rustc-dev-guide.rust-lang.org/mir/index.html#mir-constant-values that's not converted to a literal becomes one of these "const"s.

Reply:
The RHSs of those examples are all `Operand::Constant`s, and MIR only allows taking the address of a place, not a constant.  rustc handles string literals and such by creating an anonymous static allocation and emitting an `Operand::Constant` whose value is a pointer to that allocation.  It handles accesses to normal static items similarly - take a look at the MIR for [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1629be21656514623d3d3cc4769915f7).

As far as I know, any time you have a constant of pointer/reference type, it must either not point to anything (`123 as *const u8`, `&()`) or must be a pointer to a static allocation (possibly anonymous).  That's why I suggested `AddrOfStatic` as a possible name for this variant.

Reply:
> The RHSs of those examples are all `Operand::Constant`s, and MIR only allows taking the address of a place, not a constant. rustc handles string literals and such by creating an anonymous static allocation and emitting an `Operand::Constant` whose value is a pointer to that allocation. It handles accesses to normal static items similarly - take a look at the MIR for [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1629be21656514623d3d3cc4769915f7).

Right, but in that example even the `static` is initialized in MIR to a `const`:
```
static X: i32 = {
    let mut _0: i32;

    bb0: {
        _0 = const 123_i32;
        return;
    }
}
```

The static itself is initialized to the value of a `const` pointer.
 
> As far as I know, any time you have a constant of pointer/reference type, it must either not point to anything (`123 as *const u8`, `&()`) or must be a pointer to a static allocation (possibly anonymous). That's why I suggested `AddrOfStatic` as a possible name for this variant.

One significant difference between constants and statics (at least at the Rust language level): static can be mutable. I think it makes more sense to use `AddrOfConst` for anything that's an immutable constant (even if it's internally a static allocation), and use `AddrOfStatic` for things that are explicitly Rust `static` declarations (and potentially mutable, either via `mut` or e.g. `Cell`).

I'll admit that for the purpose in this PR (determining the size of a pointee) the difference between `const` and `static` isn't too relevant, since statics also need to be `Sized`.

Reply:
I just think `AddrOfConst` is a confusing name because things declared `const` in Rust don't have addresses.  This applies to `const C: i32 = 123;`, `fn f<const N: usize>()`, and the new const expressions `const { 2 + 2 }`.  `C`, `N`, and `const { 2 + 2 }` all become `Operand::Constant`s in the MIR, and you can't take the address of an `Operand::Constant`.  (You can write `&C`, `&N`, and `&const { 2 + 2 }` in Rust, but at the MIR level, this actually stores the `Operand::Constant` to a temporary and then takes the address of that temporary.)

>The static itself is initialized to the value of a const pointer.

In this particular example, it happens to be initialized with an `Operand::Constant` of type `i32`, but that isn't always the case.  For example, `static X: i32 = 123 + 456;` produces this MIR:
```Rust
static X: i32 = {
    let mut _0: i32;
    let mut _1: (i32, bool);

    bb0: {
        _1 = AddWithOverflow(const 123_i32, const 456_i32);
        assert(!move (_1.1: bool), "attempt to compute `{} + {}`, which would overflow", const 123_i32, const 456_i32) -> [success: bb1, unwind continue];
    }

    bb1: {
        _0 = move (_1.0: i32);
        return;
    }
}
```

Accessing `X` uses a const pointer:

```Rust
    bb0: {
        _1 = const {alloc1: &i32};
        _0 = (*_1);
        return;
    }
```

Essentially what has happened here is the expression `X` was expanded to `*(&X)`, then `&X` was const-evaluated (since the address of a static is a constant) and produced `{alloc1: &i32}`.  The statement `_1 = const {alloc1: &i32};` stores the address of the static `X` into `_1`, hence my suggestion to name this operation `AddrOfStatic`.

>One significant difference between constants and statics (at least at the Rust language level): static can be mutable.

Compilation of `static mut` works roughly the same way:

```Rust
    bb0: {
        _1 = const {alloc1: *mut i32};
        _2 = _1 as *const () (PtrToPtr);
        _3 = _2 as usize (Transmute);
        _4 = AlignOf(i32);
        _5 = Sub(_4, const 1_usize);
        _6 = BitAnd(_3, _5);
        _7 = Eq(_6, const 0_usize);
        assert(_7, "misaligned pointer dereference: address must be a multiple of {} but is {}", _4, _3) -> [success: bb1, unwind unreachable];
    }

    bb1: {
        _0 = (*_1);
        return;
    }
```

Here it's turned the expression `X` into `*(ptr::addr_of_mut!(X))`, and it const-evaluated `ptr::addr_of_mut!(X)` to `{alloc1: *mut i32}`.  This is again storing the address of the static into `_1`.  (The rest of `bb0` is an alignment check for the raw-pointer dereference - I guess rustc now generates those automatically in debug mode.)

---

In the interest of not spending too much more time on this particular detail, if you still disagree about the naming, just mark this resolved and go ahead with the current one.

Reply:
Looking at the event code again, I'm thinking maybe it makes more sense to call it `AddrOfSized` anyway, since that's what it actually does. Then it could be used for constants _and_ locals.